### DISCUSSION TITLE
Support Shelly Pro 3EM

### DISCUSSION THREAD

**Discussion by Tobi266** [arendst/Tasmota#18125] [Ideas]
Hi everybody,

is the Shelly Pro 3EM already supported?
If not will it be possible to run tasmota on it?
Controller is a ESP32 DOWD Q6
Teardown: https://www.shelly-support.eu/forum/thread/18815-teardown-shelly-pro-3em-stromaufnahme/

Greetings

**Comment by barbudor** [comment]
There are search tools both in the template web site and in Github Discussions

https://templates.blakadder.com/shelly_3EM.html

https://github.com/arendst/Tasmota/discussions/13515

**Reply by Tobi266** [reply]
This is not the PRO version...

**Reply by barbudor** [reply]
Sorry, my bad
PRO is under work AFAIK

**Comment by Ramalama2** [comment]
Hey Boys, i can't wait to flash tasmota onto it either :-)
Everything is tasmotized except the Pro 3EM :-)

Cheers and Thx!

**Reply by jottr** [reply]
Why would you think that there are only "boys" involved in the Tasmota project?

**Comment by msperl** [comment]
The accessible GPIO pins are found here:
https://kb.shelly.cloud/knowledge-base/shelly-pro-3em

1: esp_debug_uart
2: U0TXD
3: U0RXD
4: +3v3
5: reset
6: GPIO0
7: GND

CAUTION! High voltage on the add-on interface when the Device is powered!

**Reply by msperl** [reply]
As the calibration url on the Shelly to dump the meter (similar to “classic” 3em to read the config for the ADE7880) is not working I am right now dumping the firmware to see if I find something of interest there…

I have to admit I have not tried to open the device (yet - the tear down down not mention how to do that without braking the case…), so I guess that I will need to find out those settings “blind”…

Onte thing seems obvious to me from looking at the photos is that that mentioned u6(but more likely u15?) 3 pin chip (from the text of the tear down) is probably a reset chip to take the system only out of reset when the voltage has stabilized…

**Reply by msperl** [reply]
So the firmware offers this for calibration:
`
{"factory":{"calib":{"hardware_checked":true,"require":true,"n_do":false,"done":true,"n_done":false,"etalon":"EC6260XXXXXX","gains":{"aigain":268358848.000000,"avgain":61962.000000,"bigain":268353264.000000,"bvgain":67488.000000,"cigain":268353488.000000,"cvgain":69311.000000,"nigain":0.000000,"apgain":268425296.000000,"bpgain":268426848.000000,"cpgain":268419312.000000,"hpgain":0.000000,"aphcal":239.000000,"bphcal":243.000000,"cphcal":239.000000}}}}
`
So it is at least in the original firmware dump - even though the data is in a different format compared to the 3em.
So some sort of translation is needed!

I now feel brave and will flash tasmota 12.4.0 to it.

**Reply by msperl** [reply]
Before that: I believe that the calibration is in a fixed place on the SPI flash, so it may be possible to extract it after flashing!

Here the hexdump:

```
00ffd000  7b 22 66 61 63 74 6f 72  79 22 3a 7b 22 63 61 6c  |{"factory":{"cal|
00ffd010  69 62 22 3a 7b 22 68 61  72 64 77 61 72 65 5f 63  |ib":{"hardware_c|
00ffd020  68 65 63 6b 65 64 22 3a  74 72 75 65 2c 22 72 65  |hecked":true,"re|
00ffd030  71 75 69 72 65 22 3a 74  72 75 65 2c 22 6e 5f 64  |quire":true,"n_d|
00ffd040  6f 22 3a 66 61 6c 73 65  2c 22 64 6f 6e 65 22 3a  |o":false,"done":|
00ffd050  74 72 75 65 2c 22 6e 5f  64 6f 6e 65 22 3a 66 61  |true,"n_done":fa|
00ffd060  6c 73 65 2c 22 65 74 61  6c 6f 6e 22 3a 22 45 43  |lse,"etalon":"EC|
```

So it seems likely that one could easily use the offset 00ffd000 to read the calibrations…
(But that needs to get confirmed on other devices)

**Reply by msperl** [reply]
And looking at the partition table dump of the original firmware during boot shows:
```
I (63) boot: Partition Table:
I (67) boot: ## Label            Usage          Type ST Offset   Length   F
I (74) boot:  0 nvs              WiFi data        01 02 00009000 00004000 0
I (82) boot:  1 otadata          OTA data         01 00 0000d000 00002000 1
I (90) boot:  2 app_0            OTA app          00 10 00010000 00200000 1
I (97) boot:  3 fs_0             FS               01 82 00210000 00080000 1
I (105) boot:  4 app_1            OTA app          00 11 00290000 00200000 1
I (113) boot:  5 fs_1             FS               01 82 00490000 00080000 1
I (120) boot:  6 emdata           FS               01 82 00510000 00aec000 0
I (128) boot:  7 shelly           Unknown data     01 88 00ffc000 00004000 1
I (136) boot: End of partition table
```
(And yes: at 00ffc000 there is a json with the cloud key of the device…)

**Reply by msperl** [reply]
OK, so Ethernet is working with this extremely basic template, that now supports ethernet (taken from Shelly Pro 2)
```
{"NAME":"Shelly Pro 3EM","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,5600,6214,0,1,1,5568,0,1,1,1,0,0,0,0,1,1,1,1,1,0,0,1],"FLAG":0,"BASE":1}
```

**Reply by sfromis** [reply]
This template is "less basic", but I'm not sure how much of the energy monitoring really works.
```
{"NAME":"Shelly Pro 3EM","GPIO":[1,8736,1,1,8705,1,1,1,1,608,1,640,8704,1,5600,1,0,1,1,5568,0,1,1,1,0,0,0,0,1,1,1,32,4736,0,0,1],"FLAG":0,"BASE":1}
```

**Reply by msperl** [reply]
which firmware does that require? tasmota32 - 12.4.0 or a dev build?

**Reply by sfromis** [reply]
Don't know. The newer the better..... Maybe the calibration stuff is needed too, I've not got the device.

Of course, without an official template, the device is not "supported" as of now.

**Reply by msperl** [reply]
Well - I tried to switch to Dev Build but found myself in an ugly situation:
used the WRONG URL and tasmota used all its configuration in safeboot - but then there is no way to change wifi config, etc...
This on a DIN RACK is not as easy to recover from...

But back to basics: recovering worked and with the dev build I do see the ADE7880 and some values - even if they are 0 for now (I guess I need to make the device self powered and attach one of those sensors again!

**Reply by sfromis** [reply]
If by "safe mode" you mean safeboot, that's just an indication that you need to complete step 2 of the upgrade, and recovery is simple. Try issuing the command `Upgrade 1`. To reduce the risk of that stoppage, it can be useful to upgrade the safeboot firmware too.

**Reply by msperl** [reply]
well - for some reason I was in AP mode in safeboot - so that becomes a bit hard to change the wifi settings there...
But OK, I am flashing everything from the dev build now!

**Reply by msperl** [reply]
OK, so it sems to be working - but it is somewhat fragile!
Sometimes the Device does not find the ADE7880 after a reboot (i.e from the UI) and then it requires a power off (not very practical when hardwired on a DIN RAIL...)

**Reply by msperl** [reply]
I guess I need to test disabling serial logging to see if that solves that issue…

**Reply by msperl** [reply]
Anyway: thanks for providing that template! Any Ideas about the LED?
Any Idea how to translate the calibration from Shelly to Tasmota?

**Reply by msperl** [reply]
OK - serial logging is disabled (automatically - possibly because of my automation - not sure...), so it should not trigger issues with I2C...

Which is still used and the device is detected:
```
00:00:00.327 BRY: Berry initialized, RAM used=4041 bytes
00:00:00.343 BRY: No 'preinit.be'
00:00:00.394 CFG: Energy loaded from file
00:00:00.396 I2C: ADE7880 found at 0x38
00:00:00.548 SRC: Restart
```
so energy (taken from real file - not rule) is loaded.
I2C device ADE7880 is Detected at i2c 0x38.

Unfortunately there is nothing obvious in the logs (4 - more debug) that I can see that would indicate an issue when rebooting the device...
And I do not see an obvious difference in the logs between reboot and power on (and I have forwarded logs to syslog just to get most data...)

**Reply by slimcdk** [reply]
I've got my hands on multiple units. A few ones are early access (prototypes) and factory calibrated for 400A. I can confirm that all samples I have (that are not factory calibrated) contain the same data structure at `00ffd000`. However, the factory calibrated ones I have, have additional calibration data at `00ffc000`

```json
{
    "factory": {
        "version": 2,
        "model": <removed>,
        "batch": "dev-001",
        "cloud_key": <removed>,
        "calib": {
            "require": true,
            "n_do": false,
            "gains": {
                "avgain": 6710,
                "bvgain": 6710,
                "cvgain": 10066,
                "aigain": 268353248,
                "bigain": 268353248,
                "cigain": 268354506,
                "nigain": 268353248,
                "apgain": 268430423,
                "bpgain": 268430423,
                "cpgain": 268432101,
                "aphcal": 153,
                "bphcal": 153,
                "cphcal": 145
            }
        },
        "ade7880_x": {
            "voltage_scale_a": 7.26e-05,
            "voltage_scale_b": 7.26e-05,
            "voltage_scale_c": 7.26e-05,
            "current_scale_a": 0.000106197359,
            "current_scale_b": 0.000106197359,
            "current_scale_c": 0.000106197359,
            "current_scale_n": 0.000106197359,
            "apower_scale_a": 0.025248528,
            "apower_scale_b": 0.025248528,
            "apower_scale_c": 0.025248528,
            "aprtpower_scale_a": 0.025248528,
            "aprtpower_scale_b": 0.025248528,
            "aprtpower_scale_c": 0.025248528,
            "aenergy_scale_a": 0.000569955242,
            "aenergy_scale_b": 0.000569955242,
            "aenergy_scale_c": 0.000569955242,
            "aprtenergy_scale_a": 0.000569955242,
            "aprtenergy_scale_b": 0.000569955242,
            "aprtenergy_scale_c": 0.000569955242,
            "faenergy_scale_a": 0.000569955242,
            "faenergy_scale_b": 0.000569955242,
            "faenergy_scale_c": 0.000569955242,
            "frenergy_scale_a": 0.000569955242,
            "frenergy_scale_b": 0.000569955242,
            "frenergy_scale_c": 0.000569955242
        }
    }
}
```

**Reply by slimcdk** [reply]
I have not yet tried to calibrate any of the units. I'll try flashing a non-calibrated on with Tasmota this week.

**Comment by msperl** [comment]
As for rebooting: I did now enable Debugging (`#define ADE7880_DEBUG`) based on the latest development branch

and here the filtered output:

On initial power on:
```
ESP-I2C: ADE7880 found at 0x38u
ESP-A78: Rd 0xE503 0x01A08007 (27295751)u
ESP-A78: Wr 0xEC01 0x02 (2)u
ESP-A78: Wr 0xE503 0x3FFE8930 (1073645872)u
ESP-A78: Rd 0xE503 0x01A00007 (27262983)u
ESP-A78: Rd 0xE707 0x01 (1)u
ESP-A78: Wr 0xE60F 0x0000 (0)u
ESP-A78: Wr 0x4381 0xFFF44AE2 (-767262)u
ESP-A78: Wr 0x4380 0x003050B1 (3166385)u
ESP-A78: Wr 0x4389 0xFFEB76E4 (-1345820)u
ESP-A78: Wr 0xE614 0x00B4 (180)u
ESP-A78: Wr 0x4383 0xFFF459D1 (-763439)u
ESP-A78: Wr 0x4382 0x002FB1BB (3125691)u
ESP-A78: Wr 0x438B 0xFFEB7100 (-1347328)u
ESP-A78: Wr 0xE615 0x00B0 (176)u
ESP-A78: Wr 0x4385 0xFFF48EE2 (-749854)u
ESP-A78: Wr 0x4384 0x002FCA4F (3131983)u
ESP-A78: Wr 0x438D 0xFFEB5ED5 (-1351979)u
ESP-A78: Wr 0xE616 0x00B0 (176)u
ESP-A78: Wr 0x4386 0x001ACD8D (1756557)u
ESP-A78: Wr 0x4386 0x001ACD8D (1756557)u
ESP-A78: Wr 0x4386 0x001ACD8D (1756557)u
ESP-A78: Rd 0x4381 0x0FF44AE2 (267668194)u
ESP-A78: Rd 0x4380 0x003050B1 (3166385)u
ESP-A78: Rd 0x4389 0x0FEB76E4 (267089636)u
ESP-A78: Rd 0xE614 0x00B4 (180)u
ESP-A78: Rd 0x4383 0x0FF459D1 (267672017)u
ESP-A78: Rd 0x4382 0x002FB1BB (3125691)u
ESP-A78: Rd 0x438B 0x0FEB7100 (267088128)u
ESP-A78: Rd 0xE615 0x00B0 (176)u
ESP-A78: Rd 0x4385 0x0FF48EE2 (267685602)u
ESP-A78: Rd 0x4384 0x002FCA4F (3131983)u
ESP-A78: Rd 0x438D 0x0FEB5ED5 (267083477)u
ESP-A78: Rd 0xE616 0x00B0 (176)u
ESP-A78: Rd 0x4386 0x001ACD8D (1756557)u
ESP-A78: Wr 0xE702 0x09 (9)u
ESP-A78: Wr 0xE60C 0x0064 (100)u
ESP-A78: Wr 0xE50A 0x00000020 (32)u
ESP-A78: Wr 0xE50A 0x00000020 (32)u
ESP-A78: Wr 0xE7FE 0xAD (173)u
ESP-A78: Wr 0xE7E3 0x80 (128)u
ESP-A78: Wr 0xE228 0x0201 (513)u
ESP-A78: Rd 0xE502 0x000BFD20 (785696)u
ESP-A78: Wr 0xE502 0x00000020 (32)u
ESP-A78: Rd 0xE502 0x000BFD00 (785664)u
ESP-A78: Rd 0xE502 0x000BFD20 (785696)u
ESP-A78: Wr 0xE502 0x00000020 (32)u
ESP-A78: Rd 0xE502 0x000BFD00 (785664)u
ESP-A78: Rd 0xE502 0x000BFD20 (785696)u
ESP-A78: Wr 0xE502 0x00000020 (32)u
ESP-A78: Rd 0xE502 0x000BFD00 (785664)u
ESP-A78: Rd 0x43C6 0x000004D9 (1241)u
ESP-A78: Rd 0x43C1 0x000003A9 (937)u
ESP-A78: Rd 0x43C0 0x00000585 (1413)u
ESP-A78: Rd 0xE513 0xFFFFFFFF (-1)u
ESP-A78: Rd 0xE519 0x00000000 (0)u
ESP-A78: Rd 0xE905 0x04F5 (1269)u
ESP-A78: Rd 0xE400 0x00000000 (0)u
ESP-A78: Rd 0x43C3 0x000003A8 (936)u
ESP-A78: Rd 0x43C2 0x0000057A (1402)u
ESP-A78: Rd 0xE514 0xFFFFFFFF (-1)u
ESP-A78: Rd 0xE51A 0x00000000 (0)u
ESP-A78: Rd 0xE906 0x1451 (5201)u
ESP-A78: Rd 0xE401 0x00000000 (0)u
ESP-A78: Rd 0x43C5 0x002C53C5 (2905029)u
ESP-A78: Rd 0x43C4 0x000005F3 (1523)u
ESP-A78: Rd 0xE515 0x00000055 (85)u
ESP-A78: Rd 0xE51B 0x000001BA (442)u
ESP-A78: Rd 0xE907 0x13FD (5117)u
ESP-A78: Rd 0xE402 0x00000000 (0)u
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
ESP-A78: Reinit
```

and after a reboot:
```
ESP-I2C: ADE7880 found at 0x38
```
so there is nothing that is found besides the device on i2c!
The driver does not even initialize - or so it seems

**Comment by msperl** [comment]
so I have added a debug for ininit:

```
diff --git a/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino b/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino
index 08fb4a11f..7c99ec44a 100644
--- a/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino
+++ b/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino
@@ -367,6 +367,7 @@ int32_t Ade7880ReadVerify(uint16_t reg) {
 /*********************************************************************************************/

 bool Ade7880Init(void) {
+  AddLog(LOG_LEVEL_DEBUG, PSTR("A78: InInit"));
   // Init sequence about 6mS after reset - See page 40 (takes 68ms)
 #ifdef ADE7880_PROFILING
   uint32_t start = millis();
```

And repowered the device (now taken from web console):
```
00:00:00.315 BRY: No 'preinit.be'
00:00:00.362 CFG: Energy loaded from file
00:00:00.363 I2C: ADE7880 found at 0x38
00:00:00.412 A78: InInit
00:00:00.414 A78: Rd 0xE503 0x01A08007 (27295751)
00:00:00.416 A78: Wr 0xEC01 0x02 (2)
00:00:00.418 A78: Wr 0xE503 0x3FFE8930 (1073645872)
00:00:00.421 A78: Rd 0xE503 0x01A00007 (27262983)
...
00:00:00.506 A78: Wr 0xE7E3 0x80 (128)
00:00:00.507 A78: Wr 0xE228 0x0201 (513)
00:00:00.509 NRG: Init driver 23
00:00:00.512 SRC: Restart
00:00:00.517 Project tasmota - nit-t-k-herd Version 12.4.0.4(tasmota32)-2_0_7(2023-03-30T17:11:29)
```

and on `restart 1`:
```
00:00:00.384 BRY: No 'preinit.be'
00:00:00.447 CFG: Energy loaded from file
00:00:00.448 I2C: ADE7880 found at 0x38
00:00:00.616 SRC: Restart
00:00:00.621 Project tasmota - nit-t-k-herd Version 12.4.0.4(tasmota32)-2_0_7(2023-03-30T17:11:29)
```
(note that I had to increase the line buffer as well to get those early log lines)

So in the end: for some reason on power up i2c detects the device and then the Driver init (Ade7880Init) is called.
while on restart i2c still detects the device, but Ade7880Init is not called.

I do not have a deep understanding of the initialization process in tasmota, so someone with more knowledge in this area needs to look into this (or at least give a good hint where to look exactly, so that more diagnostics can get added)...

**Comment by msperl** [comment]
ok, so I found something else by instrumenting the dispatcher function:

PowerOn:
```
00:00:00.363 BRY: No 'preinit.be'
00:00:00.417 CFG: Energy loaded from file
00:00:00.418 A78: 1-Xnrg23 - function 3
00:00:00.418 A78: 1-i2C enabled
00:00:00.419 I2C: ADE7880 found at 0x38
00:00:00.475 A78: InInit
00:00:00.477 A78: Rd 0xE503 0x01A08007 (27295751)
00:00:00.479 A78: Wr 0xEC01 0x02 (2)
00:00:00.481 A78: Wr 0xE503 0x3FFE8930 (1073645872)
```
On restart:
```
00:00:00.273 BRY: No 'preinit.be'
00:00:00.310 CFG: Energy loaded from file
00:00:00.311 A78: 1-Xnrg23 - function 3
00:00:00.312 A78: 1-i2C enabled
00:00:00.312 I2C: ADE7880 found at 0x38
00:00:00.452 A78: Xnrg23 - function 3 result=0
00:00:00.455 SRC: Restart
00:00:00.459 Project tasmota - nit-t-k-herd Version 12.4.0.4(tasmota32)-2_0_7(2023-03-30T19:49:56)
```

At a loss where it is going wrong...

**Reply by sfromis** [reply]
Yeah, there were reasons why I found a template, while still warning that this did not mean "supported" :grinning:

**Reply by msperl** [reply]
well - As far as I can tell (instrumenting the code further) the reset logic with the RESET GPIO is not working as expected!

The Interrupt line here never goes low: https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino#L463

PowerOn (Working):
```
00:00:00.152 BRY: No 'preinit.be'
00:00:00.169 CFG: Energy loaded from file
00:00:00.170 I2C: ADE7880 found at 0x38
00:00:00.175 TFS: File 'calib.dat' not found
00:00:00.178 A78: SetCalibrate - post reset 1
00:00:00.179 A78: SetCalibrate - found
00:00:00.179 A78: InInit
```

Restart 1 (Not working):
```
00:00:00.340 CFG: Energy loaded from file
00:00:00.342 I2C: ADE7880 found at 0x38
00:00:00.350 TFS: File 'calib.dat' not found
00:00:00.352 A78: SetCalibrate - post reset 1
00:00:00.453 A78: SetCalibrate - not found
00:00:00.455 SRC: Restart
```

Is maybe the Reset Pin not identified correctly?

I guess a message in the code would help identify such a situation in the future...

For context here the instrumentation patch:
```
diff --git a/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino b/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino
index 08fb4a11f..ee1c6b569 100644
--- a/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino
+++ b/tasmota/tasmota_xnrg_energy/xnrg_23_ade7880.ino
@@ -367,6 +367,7 @@ int32_t Ade7880ReadVerify(uint16_t reg) {
 /*********************************************************************************************/

 bool Ade7880Init(void) {
+  AddLog(LOG_LEVEL_DEBUG, PSTR("A78: InInit"));
   // Init sequence about 6mS after reset - See page 40 (takes 68ms)
 #ifdef ADE7880_PROFILING
   uint32_t start = millis();
@@ -458,9 +459,12 @@ bool Ade7880SetCalibrate(void) {

   Ade7880.cycle_count = 2;                                                 // Skip first two cycles

+  AddLog(LOG_LEVEL_DEBUG, PSTR("A78: SetCalibrate - post reset %d"), pin_reset);
+
   uint32_t timeout = millis() + 100;                                       // Should be reset within 10 ms
   while (!TimeReached(timeout)) {                                          // Wait up to 100 ms
     if (!digitalRead(Pin(GPIO_ADE7880_IRQ, 1))) {
+      AddLog(LOG_LEVEL_DEBUG, PSTR("A78: SetCalibrate - found"));

 #ifdef ADE7880_PROFILING
       AddLog(LOG_LEVEL_DEBUG, PSTR("A78: Reset in %d ms"), millis() - start);
@@ -472,6 +476,7 @@ bool Ade7880SetCalibrate(void) {
       }
     }
   }
+  AddLog(LOG_LEVEL_DEBUG, PSTR("A78: SetCalibrate - not found"));
   return false;
 }
```

**Reply by msperl** [reply]
And I have to admit: I am not sure if I can open that expensive device without breaking the case rendering it unusable...
(otherwise I would trace the reset line myself...)

**Reply by msperl** [reply]
So I got brave/stupid and opened it up (4 clamps - 2 at the bottom/DIN Rail and 2 at the top).
As far as I can tell right now the RESET pin is not connected to anything except for the pullup and the capacitor!
(to the contrary to my original posting)

I shall try tracing the LED as well - (the Resistors R45-47 are routed via several vias to the ESP - details to follow)

**Reply by msperl** [reply]
in the meantime: the RTC chip is a AIP8563!

**Reply by msperl** [reply]
So I was too brave! The device is no longer 100% happy - wifi does not work any longer...

But I got a few pointers:
* WIFI Led - Blue - GPIO 32
* WIFI Led - Red - GPIO 33
* COUNT LED - possibly on ADE7880 pin 33 - CF1
* ETH LED - managed by SMSC 8720A (Ethernet) as far as I can see...

And as said: my best guess is now that there is no GPIO for reset of the ADE7880

**Reply by ramiws** [reply]
So it works on first boot but not on reset? Why is the reset of the ADE7880 needed?

**Comment by une18** [comment]
Any updates?

**Comment by une18** [comment]
Shelly launced  Pro 3EM-400  https://www.shelly.cloud/en-ee/products/product-overview/pro-3-em/shelly-pro-3-em-400

**Comment by Dacesilian** [comment]
Hello, please are there any update for support Shelly Pro 3EM? I would love to use smaller 50A transformers, so need to use alternative firmware (Tasmota) and calibrate manually. Thanks.

**Comment by gbomacfly** [comment]
Hi! Are there any news? I want to buy one of these :)

**Comment by Noschvie** [comment]
Hello,
are there any news? Thanks

**Comment by Dirtylobstar** [comment]
Any update? It was quite a while something was written here. 
I made the mistake of flashing my Shelly Pro 3em to tasmota since I didnt read the text good enough and now it is more or less unusable. Shelly Pro 3em is in the list and it boot fine but once started it actually says Shelly 3EM and nothing works since the hardware is so different. I do not have a valid file so I can revert to Shelly either because they started to be grumpy and protects the files with OTA only (my interpretation, please correct me :))
I can also use it in esphome but that it not any better since I cannot get it to work there. Chatgpt suggest code that it accept but there are no output for any values....

**Comment by goodkiller** [comment]
@arendst any updates? As we know then old 3EM is discontinued and Shelly producing only PRO version.

**Comment by therapys** [comment]
Hey, It's been a while since this got a reply again, is there any possiblity this will be done in the near future?

**Comment by inox100** [comment]
Hallo, ich habe den Shelly Pro 3EM jetzt mit Tasmota am laufen. Er überwacht 3 Inverter. Es gibt aber noch ein paar Probleme. Aus dem Kaltstart wird der Treiber für den ADE7880 geladen, wenn ich einen Software-Start mache, fehlen die beiden Zeile "Energy loaded" und "Init driver 23" und es werden keine Werte angezeigt, nur die Temperatur.

Auszug aus der Konsole.
Kaltstart
00:00:00.089 BRY: Berry initialized, RAM used 3796 bytes
00:00:00.096 I2C: ADE7880 found at 0x38
00:00:00.188 CFG: Energy loaded from file
00:00:00.188 NRG: Init driver 23
00:00:00.193 Projekt tasmota - 2000W - 3000W - 3500W Version 14.6.0(release-tasmota32)-3_1_3(2025.04.16 11:28:37
Warmstart
00:00:00.079 BRY: Berry initialized, RAM used 3796 bytes
00:00:00.085 I2C: ADE7880 found at 0x38
00:00:00.192 Projekt tasmota - 2000W - 3000W - 3500W Version 14.6.0(release-tasmota32)-3_1_3(2025.04.16 11:28:37)

Hat jemand eine Idee, wie ich das Problem lösen könnte?

Es gibt noch eins. Die Anzeige friert immer mal ein. Einmal am Tag oder öfters. Alle aktuellen Werte stehen auf Null, nur die gespeicherten Werte wie Energie Gestern oder gesamt sind noch die alten.
Grüße

**Reply by Thalhammer** [reply]
Hi, I think I got pretty similar rn. I have the device flashed with tasmota and the simple things like led's, ethernet and wifi work fine for me.
I also extracted my calibration data (I dumped the whole flash just in case) and converted it to the format that tasmota expects (note that the json keys are different).

However I am getting bogus values. According to tasmota my 230V power outlet delivers 286V without calibration data and 319V when I apply the calibration data from the original firmware. I experience the same issue with the ADE going seemingly offline after a software reboot, but I suspect thats just a bug in the driver for it.

**Comment by WaDeSo** [comment]
I got a Pro 3EM running reliably on custom ESP-IDF firmware and documented everything, so here are answers to the open questions in this thread:

**ADE7880 dead after a software restart** – the ADE7880 reset pin is not connected to the ESP32 (as msperl found). After an ESP32-only restart the chip keeps running with its old setup: DSP started, DSP memory write-protected, RSTDONE already acknowledged. A driver waiting for a fresh RSTDONE never gets one. What works: `RUN=0`, open write protection (`0xAD`→`0xE7FE`, `0x00`→`0xE7E3`), software reset (`CONFIG` bit 7), wait for RSTDONE in `STATUS1`, then write **all** calibration registers including zeros, read them back, protect, `RUN=1`. Verified across repeated OTA restarts.

**Calibration** – there is no `calib.dat` on the Pro. It is JSON in the flash partition `shelly` (subtype 0x88) at 0xFFC000, calibration block at +0x1000. The gains are raw 24-bit values sign-extended to 28 bits (e.g. `268328160` = 0x0FFE5EE0 = −106784 ≈ −1.3 %); read as plain 32-bit integers they give very wrong results. Also note the block only contains gains – the counts→V/A/W scales are different from the old 3EM, so the 3EM constants will not give correct volts.

**Going back to stock** – only with a full flash dump taken *before* flashing (`esptool read-flash 0 ALL` + `verify-flash`).

**Pins** – I²C SDA 15 / SCL 13 (ADE7880 0x38, RTC 0x51), Ethernet MDC 23 / MDIO 18 / RMII clock out GPIO17 / PHY 0. Bench power: 5 V at J15 works; 3.3 V at J6 resets at `phy_init` with stock firmware.

Firmware (GPL-3.0), hardware notes and board photos: https://github.com/WaDeSo/shelly-pro3em-10hz – maybe useful for a Tasmota/ESPHome driver too.
